### PR TITLE
Metrics reporter for yammer metrics -> wavefront socket.

### PR DESCRIPTION
**Pull request by moribellamy**

### CODE REVIEW

File: java-lib/src/main/java/com/wavefront/common/MetricsToTimeseries.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import com.yammer.metrics.core.Metered;
+import com.yammer.metrics.core.Sampling;
+import com.yammer.metrics.core.Summarizable;
+
+import java.util.Map;
+
+/**
+ * @author Mori Bellamy (mori@wavefront.com)
+ */
+public class MetricsToTimeseries {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricsToTimeseries.java]: utility class can be marked abstract

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Yammer MetricProcessor that sends metrics to a TCP Socket in Wavefront-format.
+ *
+ * @author Mori Bellamy (mori@wavefront.com)
+ */
+public class SocketMetricsProcessor implements MetricProcessor<Void> {
+  protected static final Logger logger = Logger.getLogger(SocketMetricsProcessor.class.getCanonicalName());
+  private String hostname;
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: final? also port and wavefrontHP

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  private int port, wavefrontHistogramPort;
+  private BufferedOutputStream metricsStream, histogramsStream;
+  private Supplier<Long> timeSupplier;
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: final?

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    resetHistogramsSocket();
+    resetMetricsSocket();
+  }
+
+  private interface Func {
+    void run() throws IOException;
+  }
+
+  private void resetMetricsSocket() throws IOException {
+    metricsStream = new BufferedOutputStream(new Socket(hostname, port).getOutputStream());
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: if this is already set, do you need to dispose? also 58

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+   */
+  private void write(String message, OutputStream stream, Func reset) throws Exception {
+    try {
+      stream.write(message.getBytes());
+      stream.flush();
+    } catch (Exception e) {
+      try {
+        reset.run();
+        logger.info("Successfully reset connection to " + hostname);
+        stream.write(message.getBytes());
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: so in this case we don't flush?

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+   * Try to send the given message. On failure, reset and try again. If _that_ fails, just rethrow the exception.
+   */
+  private void write(String message, OutputStream stream, Func reset) throws Exception {
+    try {
+      stream.write(message.getBytes());
+      stream.flush();
+    } catch (Exception e) {
+      try {
+        reset.run();
+        logger.info("Successfully reset connection to " + hostname);
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: Should we also include the exception that was caught?

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+  /**
+   * @return " k1=v1 k2=v2 ..." if metricName is an instance of TaggedMetricName. "" otherwise.
+   */
+  private String tagsForMetricName(MetricName metricName) {
+    if (metricName instanceof TaggedMetricName) {
+      TaggedMetricName taggedMetricName = (TaggedMetricName) metricName;
+      StringBuilder sb = new StringBuilder();
+      for (Map.Entry<String, String> entry : taggedMetricName.getTags().entrySet()) {
+        sb.append(" ").append(entry.getKey()).append("=").append(entry.getValue());
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: you need to quote

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+      }
+      return sb.toString();
+    } else {
+      return "";
+    }
+  }
+
+  private void writeMetric(MetricName metricName, String nameSuffix, double value) throws Exception {
+    StringBuilder sb = new StringBuilder();
+    sb.append(metricName.getName());
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/SocketMetricsProcessor.java]: quotes for metric name

File: yammer-metrics/src/main/java/com/wavefront/integrations/metrics/WavefrontYammerMetricsReporter.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  @Override
+  public void run() {
+    try {
+      for (Map.Entry<String, SortedMap<MetricName, Metric>> entry : getMetricsRegistry().groupedMetrics().entrySet()) {
+        for (Map.Entry<MetricName, Metric> subEntry : entry.getValue().entrySet()) {
+          subEntry.getValue().processWith(socketMetricProcessor, subEntry.getKey(), null);
+        }
+      }
+    } catch (Exception e) {
+      logger.log(Level.SEVERE, "Cannot report point to Wavefront, Dropping!", e);
```
Comment (panghy) [inline yammer-metrics/src/main/java/com/wavefront/integrations/metrics/WavefrontYammerMetricsReporter.java]: We should inform the user that we cannot report yammer metrics to Wavefront and will retry in another cycle

**Comment by moribellamy** [comment]
@panghy iterated PTAL

**Comment by moribellamy** [comment]
@panghy iterated ptal

**Pull request merged**